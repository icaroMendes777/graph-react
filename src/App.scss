
body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}


.padding_03{
  padding:0.3rem;
}


.App{
  width:98vw;
  padding: 10px;
  display: grid;
  grid-template-areas: 'header header'
                      'form graph';
  grid-template-columns: 250px auto;
}

.intro{
  grid-area: header;
}

.months{
  grid-area: form;
  text-transform:capitalize;

  label{
    display: grid;
    grid-template-columns: 1fr 1fr;
    input{
      width:2rem;
    }
  }
}

.form_year{
  
  display:grid;
  grid-template-columns: auto;
}

.graph{
  grid-area: graph;
  border: 1px solid black;

  display: grid;
  grid-template-columns: repeat(12, 1fr);
  align-items: end;
  max-width: 1100px;
}

.month_name_graph{
  Overflow: hidden;
}


.graph_month{
  
  &:hover{
    border:3px solid black;
  }
}


@media only screen and (max-width:800px) {
  .App{
    grid-template-areas: 'header'
                        'form '
                        'graph';
    grid-template-columns: auto;
  }
  
}

@media only screen and (max-width:600px) {


  .graph{
    width: 600px;
    //overflow-x:scroll;
  
  }
}