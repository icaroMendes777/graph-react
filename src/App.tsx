import logo from './logo.svg';
import './App.scss';

import React from 'react';


/**
 * O seguinte componente foi criado utilizando um componente de classe.
 * como componentes de classe estão saido de uso, tal aplicação foi feita 
 * com a intenção de estudo de código legado.
 */


class App extends React.Component {

  constructor(props) {
    super(props);

    this.maxHeightColumn = 400;
    

    this.yearIncome = [
      {name:'janeiro',
      value:275},
      {name:'fevereiro',
      value:120},
      {name:'março',
      value: 180},
      {name:'abril',
      value:450},
      {name:'maio',
      value:250},
      {name:'junho',
      value:500},
      {name:'julho',
      value:450},
      {name:'agosto',
      value:700},
      {name:'setembro',
      value:350},
      {name:'outubro',
      value:300},
      {name:'novembro',
      value:200 },
      {name:'dezembro',
      value:75}
    ];


    /**
     * valores teste são carregados no inicio da aplicação
     */
    
    this.total = 0;

    this.yearIncome.forEach((el,i)=>{
      
      if(el.value) this.total += parseInt(el.value);
      
    })

    this.state = {value :0,
                  yearTotalIncome: this.total,
                  yearIncome:this.yearIncome}


             
    this.arrayColors = [
      'blue', 'green', 'red', 
      'blue', 'green', 'red', 
      'blue', 'green', 'red', 
      'blue', 'green', 'red', 
    ]
  }

  componentDidMount(){
    document.title = "Gráfico App"
  }



  //===========================================================
  ///FUNCTIONS


  calculateTotalYearIncome = ()=>{


    let total=0;
    this.state.yearIncome.forEach((el,i)=>{
      
      if(el.value) total += parseInt(el.value);
        //console.log('-----',total);
    })
    
    return total;
  }

  //-------------------------------------

  calculateAverageMonth = ()=>{

    return 0;

    return Math.trunc( ( this.calculateTotalYearIncome() / 12) );
  }



  //-------------------------------------


  makeStyleColumnMonth = (monthIndex)=>{

          if(!this.yearIncome[monthIndex].value) return  ({height: '0px'});

          let monthValue = this.yearIncome[monthIndex].value;
          let highestMonthVal = this.getHighestMonthValue();


            //console.log(highestMonthVal);

         
          let heightPx = (this.maxHeightColumn * monthValue) / highestMonthVal;


          if(heightPx > 400){
            alert("!");
            console.log('hp',heightPx);
            console.log('mv',monthValue);
            console.log('hm', highestMonthVal);
            return;
          }
          heightPx = heightPx+'px';
          
          return ({
            height: heightPx,
            backgroundColor:this.arrayColors[monthIndex],
          });
  }


  //-------------------------------------

  handleIncome = (e,i)=> {
    
    if(e.target.value >9999){
      alert('valor máximo 9999');
      return;
    }

    let newState = this.state;
    
    if(e.target.value) newState.yearIncome[i].value=parseInt(e.target.value);
    else newState.yearIncome[i].value=0;

    newState.yearTotalIncome = this.calculateTotalYearIncome();

    this.setState(newState);
    
    return;
  }



  //-------------------------------------



  getHighestMonthValue = ()=>{
      let highestValue = 0;
      this.state.yearIncome.forEach((el)=>{
          let curValue = parseInt(el.value);
          if(curValue > highestValue) highestValue = curValue;
      });

      return(highestValue);
  }


  //-------------------------------------



  preventNotNumberInput = (event)=>{
    if (!/[0-9]/.test(event.key)) {
      event.preventDefault();
    }
    return;
  }



  //-------------------------------------

  truncate_3 = (words:string)=>{
    return `${words.slice(0, 3)}.`
  }


  
////============================================================

  render(){


      return (
        <div className="App">
          <div className='intro'>
            <h1>Balanço Anual:</h1>
          </div>


            <div className="months">

            <form>

            {this.state.yearIncome.map((el,i)=>{
                    return(
                      <label key={i} className='padding_03'>
                        <div className='month_name_form'>{el.name}</div>
                      <input
                              type="text"
                              name={el.name}
                              value={el.value}
                              onChange={(event)=>this.handleIncome(event,i)}
                              onBlur={this.handleBlurMonth}
                              key={i}

                              onKeyPress={(event) => {
                                this.preventNotNumberInput(event);
                              }}
                            />
                      </label>
                    )
            })}

              {
                //--------------------------------------------------------
              }

          
              </form>

            <div className='padding_03'>
            <p>Balanço total: R$<b> {this.state.yearTotalIncome},00</b></p>
            

            </div>
            </div>


            <div className='graph'>

            {this.state.yearIncome.map((el,i)=>{
                    return(

                      <div className='wrap_month Column' key={i}>
                        <div className='graph_month' title={el.name+'- R$'+el.value} style={this.makeStyleColumnMonth(i)}>
                        </div>

                        <div className='month_name_graph padding_03'>{this.truncate_3(el.name)}</div>
                      </div>
                    )
            })}



            </div>



            {
                //--------------------------------------------------------
              }


        <footer>
          <small>App desenvolvido por <a href="http://icaromendes.epizy.com/">Icaro Mendes.</a></small>
        </footer>


        </div>
      );

  }
}



export default App;
