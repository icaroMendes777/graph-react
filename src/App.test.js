import { render, screen } from '@testing-library/react';
import App from './App';
import renderer from "react-test-renderer";



describe("Jest Snapshot testing suite", () => {
  it("Matches DOM Snapshot", () => {
    const domTree = renderer.create(<App />).toJSON();
    expect(domTree).toMatchSnapshot();
  });
});


test('renders learn react link', () => {  
  render(<App />);
  const linkElement = screen.getByText(/Balan√ßo Anual/i);
  expect(linkElement).toBeInTheDocument();
});
