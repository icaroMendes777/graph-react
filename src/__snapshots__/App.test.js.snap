// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Jest Snapshot testing suite Matches DOM Snapshot 1`] = `
<div
  className="App"
>
  <div
    className="intro"
  >
    <h1>
      Balanço Anual:
    </h1>
  </div>
  <div
    className="months"
  >
    <form>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          janeiro
        </div>
        <input
          name="janeiro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={275}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          fevereiro
        </div>
        <input
          name="fevereiro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={120}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          março
        </div>
        <input
          name="março"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={180}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          abril
        </div>
        <input
          name="abril"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={450}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          maio
        </div>
        <input
          name="maio"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={250}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          junho
        </div>
        <input
          name="junho"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={500}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          julho
        </div>
        <input
          name="julho"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={450}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          agosto
        </div>
        <input
          name="agosto"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={700}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          setembro
        </div>
        <input
          name="setembro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={350}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          outubro
        </div>
        <input
          name="outubro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={300}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          novembro
        </div>
        <input
          name="novembro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={200}
        />
      </label>
      <label
        className="padding_03"
      >
        <div
          className="month_name_form"
        >
          dezembro
        </div>
        <input
          name="dezembro"
          onChange={[Function]}
          onKeyPress={[Function]}
          type="text"
          value={75}
        />
      </label>
    </form>
    <div
      className="padding_03"
    >
      <p>
        Balanço total: R$
        <b>
           
          3850
          ,00
        </b>
      </p>
    </div>
  </div>
  <div
    className="graph"
  >
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "blue",
            "height": "157.14285714285714px",
          }
        }
        title="janeiro- R$275"
      />
      <div
        className="month_name_graph padding_03"
      >
        jan.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "green",
            "height": "68.57142857142857px",
          }
        }
        title="fevereiro- R$120"
      />
      <div
        className="month_name_graph padding_03"
      >
        fev.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "red",
            "height": "102.85714285714286px",
          }
        }
        title="março- R$180"
      />
      <div
        className="month_name_graph padding_03"
      >
        mar.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "blue",
            "height": "257.14285714285717px",
          }
        }
        title="abril- R$450"
      />
      <div
        className="month_name_graph padding_03"
      >
        abr.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "green",
            "height": "142.85714285714286px",
          }
        }
        title="maio- R$250"
      />
      <div
        className="month_name_graph padding_03"
      >
        mai.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "red",
            "height": "285.7142857142857px",
          }
        }
        title="junho- R$500"
      />
      <div
        className="month_name_graph padding_03"
      >
        jun.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "blue",
            "height": "257.14285714285717px",
          }
        }
        title="julho- R$450"
      />
      <div
        className="month_name_graph padding_03"
      >
        jul.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "green",
            "height": "400px",
          }
        }
        title="agosto- R$700"
      />
      <div
        className="month_name_graph padding_03"
      >
        ago.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "red",
            "height": "200px",
          }
        }
        title="setembro- R$350"
      />
      <div
        className="month_name_graph padding_03"
      >
        set.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "blue",
            "height": "171.42857142857142px",
          }
        }
        title="outubro- R$300"
      />
      <div
        className="month_name_graph padding_03"
      >
        out.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "green",
            "height": "114.28571428571429px",
          }
        }
        title="novembro- R$200"
      />
      <div
        className="month_name_graph padding_03"
      >
        nov.
      </div>
    </div>
    <div
      className="wrap_month Column"
    >
      <div
        className="graph_month"
        style={
          Object {
            "backgroundColor": "red",
            "height": "42.857142857142854px",
          }
        }
        title="dezembro- R$75"
      />
      <div
        className="month_name_graph padding_03"
      >
        dez.
      </div>
    </div>
  </div>
  <footer>
    <small>
      App desenvolvido por 
      <a
        href="http://icaromendes.epizy.com/"
      >
        Icaro Mendes.
      </a>
    </small>
  </footer>
</div>
`;
